caffe2/core/operator_c10wrapper.h

```diff
@@ -25,104 +16,100 @@ using extract_type_t = typename ParameterDef::type;
  *
  */
 
-template <
-    const c10::OperatorHandle& (*OperatorHandle)(),
-    class Context,
-    bool use_array_input,
-    size_t num_output_parameters,
-    class ParameterDefTuple>
+namespace detail {
+template <class Context>
 class C10OperatorWrapper final : public Operator<Context> {
  public:
-  static_assert(
-      c10::guts::is_instantiation_of<std::tuple, ParameterDefTuple>::value,
-      "");
-  using ParameterTuple =
-      c10::guts::typelist::to_tuple_t<c10::guts::typelist::map_t<
-          details::extract_type_t,
-          c10::guts::typelist::from_tuple_t<ParameterDefTuple>>>;
-
   USE_OPERATOR_CONTEXT_FUNCTIONS;
 
-  C10OperatorWrapper(const OperatorDef& operator_def, Workspace* ws)
+  C10OperatorWrapper(
+      const c10::OperatorHandle& op,
+      const OperatorDef& operator_def,
+      Workspace* ws)
       : Operator<Context>(operator_def, ws),
-        op_(OperatorHandle()),
+        op_(op),
         kernel_(at::nullopt),
-        parameters_(parse_parameters_(
-            operator_def,
-            c10::guts::make_index_sequence<num_parameters()>())) {
+        has_preallocated_outputs_(
+            op_.schema().arguments().back().name() ==
+            detail::PREALLOCATED_OUTPUT_ARGNAME) {
+    AT_ASSERT(
+        !has_preallocated_outputs_ ||
+        op_.schema().arguments().back().type()->isSubtypeOf(
+            OptionalType::create(ListType::ofTensors())));
 
     AT_ASSERT(operator_def.output_size() == op_.schema().returns().size());
-    AT_ASSERT(operator_def.input_size() == num_inputs());
-  }
-
-  size_t num_inputs() {
-    return op_.schema().arguments().size() - 1 - num_parameters();
-  }
-
-  static constexpr size_t num_parameters() {
-    return std::tuple_size<ParameterDefTuple>::value;
+    AT_ASSERT(
+        operator_def.input_size() + (has_preallocated_outputs_ ? 1 : 0) <=
+        op_.schema()
+            .arguments()
+            .size()); // '<=' because there might be caffe2 arguments
```

Comment:
Would it be accurate to say that any of the trailing arguments MUST NOT be tensors?

Reply:
added it to the comment